include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,197 @@
+//
+// Copyright (c) 2022 Alan Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_IMPL_ANY_ALLOCATOR_HPP
+#define BOOST_URL_IMPL_ANY_ALLOCATOR_HPP
+
+#include <atomic>
+
+#include <boost/url/detail/over_allocator.hpp>
+
+namespace boost {
+namespace urls {
+
+// base class to store the underlying allocator bound to `base`
+template <class T>
+struct any_allocator<T>::base
+{
+  std::atomic<std::size_t> refs{1};
+
+  virtual
+  ~base()
+  {
+  }
+
+  // although the allocator in `base` is bound to `base`, `allocate` always
+  // returns `char` as make_string_value does.
+  virtual
+  char*
+  allocate(std::size_t n) = 0;
+
+  virtual
+  void
+  deallocate(char*p, std::size_t n) = 0;
+
+  template <class U>
+  void
+  deallocate(U* p, std::size_t n) {
+      deallocate(reinterpret_cast<char*>(p), n * sizeof(T));
+  }
+
+  void
+  deallocate(char*p) {
```

Comment:
why do we need this overload

Reply:
`any_allocator::deallocate` needs it because we are implementing the complete named requirement.

Reply:
I don't see that overload in the _Allocator_ requirements here https://en.cppreference.com/w/cpp/named_req/Allocator

Reply:
All allocators need to be able to deallocate. This is the formal requirement

![image](https://user-images.githubusercontent.com/5369819/149417457-f14ecc05-2c7b-4108-b422-bfddc476cc8d.png)

Reply:
The image shows a function with two parameters, while the `deallocate` function above has one parameter

Reply:
Oh... you mean that deallocate. That has been removed like 50 commits ago. I didn't even remember.